File: test/lib/c/03-publish-helper-multiple.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <mosquitto.h>

static int run = -1;

int main(int argc, char *argv[])
{
	int rc;
	struct mosquitto *mosq;
	int i = 0;

	int port = atoi(argv[1]);

	mosquitto_lib_init();

	mosq = mosquitto_new("publish-helper-multiple-qos0-test", true, NULL);
	if(mosq == NULL){
		return 1;
	}

	struct mosquitto_message messages[5];
	struct mosquitto_message *message_ptrs[5];
	char buf[50];

	for (i = 0; i < 5; i++) {
		struct mosquitto_message *m = &messages[i];
		sprintf(buf, "publish-helper-multiple/qos0/test/%d", i);
		m->topic = strdup(buf);
		m->payload = "message";
		m->payloadlen = strlen("message");
		m->qos = 0;
		m->retain = 0;

		message_ptrs[i] = m;
	}

	rc = mosquitto_publish_multiple(
			message_ptrs,
			5,
			"localhost", port,
			"publish-helper-multiple-qos0-test", 60, true,
			NULL, NULL, NULL, NULL);

	mosquitto_lib_cleanup();
	return rc;
}

File: test/lib/03-publish-helper-multiple.py
#!/usr/bin/env python3

# Test whether a client sends a correct PUBLISH to a topic with QoS 0, using
# the mosquitto_publish_multiple helper function.

# The client should connect to port 1888 with keepalive=60, clean session set,
# and client id publish-helper-multiple-qos0-test
# The test will send a CONNACK message to the client with rc=0. Upon receiving
# the CONNACK and verifying that rc=0, the client should send five PUBLISH messages
# to topic "publish-helper-multiple/qos0/test/$i" with payload "message" and QoS=0.
# If rc!=0, the client should exit with an error.
# After sending the PUBLISH messages, the client should send a DISCONNECT message.

from mosq_test_helper import *

port = mosq_test.get_lib_port()

rc = 1
keepalive = 60
connect_packet = mosq_test.gen_connect("publish-helper-multiple-qos0-test", keepalive=keepalive, clean_session=True)
connack_packet = mosq_test.gen_connack(rc=0)

publish_packets = []
for i in range(0, 5):
    publish_packets.append(mosq_test.gen_publish("publish-helper-multiple/qos0/test/%d" % i, qos=0, payload="message"))

disconnect_packet = mosq_test.gen_disconnect()

sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
sock.settimeout(10)
sock.bind(('', port))
sock.listen(5)

client_args = sys.argv[1:]
env = dict(os.environ)
env['LD_LIBRARY_PATH'] = '../../lib:../../lib/cpp'
try:
    pp = env['PYTHONPATH']
except KeyError:
    pp = ''
env['PYTHONPATH'] = '../../lib/python:'+pp
client = mosq_test.start_client(filename=sys.argv[1].replace('/', '-'), cmd=client_args, env=env, port=port)

try:
    (conn, address) = sock.accept()
    conn.settimeout(10)

    mosq_test.do_receive_send(conn, connect_packet, connack_packet, "connect")

    for i in range(0, 5):
        mosq_test.expect_packet(conn, "publish", publish_packets[i])

    mosq_test.expect_packet(conn, "disconnect", disconnect_packet)
    rc = 0

    conn.close()
except mosq_test.TestError:
    pass
finally:
    client.terminate()
    client.wait()
    if rc:
        (stdo, stde) = client.communicate()
        print(stde)
    sock.close()

exit(rc)
